### DISCUSSION TITLE
Controlling HoldBrowser behavior

### DISCUSSION THREAD

**Discussion by mark-summerfield** [fltk-rs/fltk-rs#1096] [Q&A]
I have a HoldBrowser with lots of rows. When I press `Enter` on a row the row is _de_-selected and the browser scrolls to and selects the first row.
This isn't what I want. Instead I want to have a function called when I press `Enter`.
I tried setting up a callback trigger for `EnterKey` but unfortunately (1) this didn't change the behavior described above, and (2) the callback was triggered both on pressing `Enter` and on simply navigating to a new row using the up and down arrow keys.

**Comment by MoAlyousef** [comment]
I tried setting the trigger to Enter and it seems browsers don't respond to Enter.
I think the solution is to handle the enter key in the handle method:
```rust
use fltk::{prelude::*, *};

fn main() {
    let app = app::App::default().with_scheme(app::Scheme::Gtk);

    let mut win = window::Window::default().with_size(900, 300);
    let mut b = browser::HoldBrowser::default()
        .with_size(900 - 10, 300 - 10)
        .center_of(&win);
    let widths = &[50, 100, 100];

    b.set_column_widths(widths);
    b.set_column_char('|');

    for i in 0..200 {
        b.add(&format!("value|{}|Description", i)); 
    }

    b.handle(|b, ev| match ev {
        enums::Event::KeyUp => {
            if app::event_key() == enums::Key::Enter {
                println!("{}", b.text(b.value()).unwrap());
                true
            } else {
                false
            }
        }
        _ => false,
    });

    win.end();
    win.show();
    app.run().unwrap();
}
```

Although I'm not seeing the row being deselected when pressing Enter. Do you have a repro?

**Reply by mark-summerfield** [reply]
I've tried that and yes I can get the row, but it still returns the focus to the first row even if in the handler I set it to the value()'th row and even if I use a timer to do that after 50ms.

Also, BTW, your code for setting widths _only_ works in `main()`; it won't work if you refactor it into another function.

**Reply by MoAlyousef** [reply]
I’ll check it out.

**Reply by MoAlyousef** [reply]
I've changed the way set_column_widths work in the current master branch. I still can't seem reproduce the issue where the focus returns to the first row.

**Reply by mark-summerfield** [reply]
I just retried on windows. Not only does Enter move the focus to the first item, but the window won't resize!
I've attached a tarball of the app in its current incomplete form. It builds & runs on windows & linux so you can see the issues.
[charfind.tar.gz](https://github.com/fltk-rs/fltk-rs/files/7834807/charfind.tar.gz)

I now know why it won't resize: there's a bug in `fltk::window::Window::set_size_range()`. It works great on Linux but fails on windows. I suspect because I use it purely to set a sensible minimum size and leave the max open `i32::MAX`; my guess is that this is too big for windows. If my guess is right it ought to be clamped to whatever the Windows limit is.

Anyway for now I'll use:
```rust
    main_window.size_range(
         WINDOW_WIDTH_MIN,
         WINDOW_HEIGHT_MIN,
         fltk::app::screen_size().0 as i32,
         fltk::app::screen_size().1 as i32,
     );
```

**Reply by MoAlyousef** [reply]
I added a println to check the action sent and it seems that pressing Enter actually sends an Action::Search which clears the browser first. It seems the InputChoice is always being triggered by the Enter key for some reason.
It's not default HoldBrowser behavior to move focus to the first item on pressing Enter.

**Reply by mark-summerfield** [reply]
Sorry, my mistake!

I no longer care about Enter for the input choice since I search as I type.
However, I still can't figure out how to detect Enter pressed on a browser row. Anyway, I'll let you know when I get further... Thanks for your help.

Right now the results are refreshed if you type in the input choice or press F2 (which pops up the menu) and choose a previous search or click the popup menu button and press F5. All of which works pretty well: just a pity that when you choose by mouse you have to press F5 after - wish that was done automatically like it does when pressing F2, but I don't know how.

However for the browser I can detect the user pressing Space -- but not Enter (at least not reliably i.e., when the browser has focus and it isn't just an arrow or page up/down keypress).

Here's where I'm up to: 
[charfind.tar.gz](https://github.com/fltk-rs/fltk-rs/files/7835060/charfind.tar.gz)
